external/llvm-project/mlir/lib/Conversion/AsyncToGPU/AsyncToGPU.cpp

```diff
@@ -0,0 +1,354 @@
+      return op.emitOpError("requires a gpu target");
+
+    auto arch = gpuAttr->get("arch");
+    auto binary = gpuAttr->get("binary");
+    auto blockSize = gpuAttr->get("block_size").cast<IntegerAttr>();
+    auto gridSize = gpuAttr->get("grid_size").cast<IntegerAttr>();
+
+    auto func = *getCalledFunc(op);
+    auto floc = func.getLoc();
+
+    // Also capture the accessMap for the params, default all to read-write
+    SmallVector<func::AccessMode> accessVec(func.getNumArguments(),
+                                            func::AccessMode::ReadWrite);
+    if (auto accessMap = func->getAttrOfType<ArrayAttr>("access_map")) {
+      accessVec =
+          llvm::to_vector<4>(llvm::map_range(accessMap, [](Attribute a) {
+            return a.cast<func::AccessModeAttr>().getValue();
+          }));
+    }
+
+    // 2. create dummy gpu.module for reference from gpu.launch_func
+    //    - with gpu.binary, arch attributes
+    //    - and gpu.func (referenced by gpu.launch_func
+    //    gpu.module @<func_name>_module attributes {arch = "gfx908", gpu.binary
+    //        = "\7FELF\..."} {
+    //      gpu.func @<func_name> (...) attributes {block_size = 256 : i32,
+    //          grid_size = 900 : i32, gpu.kernel}
+
+    FunctionOpInterface funcIF(func);
+    auto funcName = funcIF.getName();
+    auto gpuModuleName = funcName + "_module";
+
+    auto gpuModule = module.lookupSymbol<gpu::GPUModuleOp>(gpuModuleName.str());
+    if (!gpuModule) {
+      OpBuilder b(ctx);
+      gpuModule = b.create<gpu::GPUModuleOp>(floc, gpuModuleName.str());
+      gpuModule->setAttr("arch", arch);
+      gpuModule->setAttr("gpu.binary", binary);
+
+      SymbolTable symbolTable(module);
+      symbolTable.insert(gpuModule);
+    }
+
+    auto gpuFunc = gpuModule.lookupSymbol<gpu::GPUFuncOp>(funcName);
+    if (!gpuFunc) {
+      OpBuilder b(gpuModule.getContext());
+      gpuFunc =
+          b.create<gpu::GPUFuncOp>(floc, funcName, func.getFunctionType());
+      gpuFunc->setAttr("block_size", blockSize);
+      gpuFunc->setAttr("grid_size", gridSize);
+      gpuFunc->setAttr(gpu::GPUDialect::getKernelFuncAttrName(),
+                       b.getUnitAttr());
+
+      SymbolTable symbolTable(gpuModule);
+      symbolTable.insert(gpuFunc);
+
+      // Must have a return
+      auto block = &gpuFunc.front();
+      b.setInsertionPoint(block, block->begin());
+      b.create<gpu::ReturnOp>(floc, ValueRange{});
+    }
+
+    // 3. create substitute gpu.launch_func
+    //    %15 = gpu.wait async
+    //    %16 = gpu.launch_func async [%15] @test_fusion_module::@test_fusion
+    //    blocks in (%c900, %c1, %c1) threads in (%c256, %c1, %c1)
+    //    dynamic_shared_memory_size %c0_i32 args(%4 : memref<128x32x32x8xf32>,
+    //    %9 : memref<128x3x3x8xf32>, %14 : memref<128x30x30x128xf32>)
+
+    auto tokenType = rewriter.getType<gpu::AsyncTokenType>();
+
+    auto zeroIdx =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(0));
```

Comment:
`createOrFold`, perhaps?
And sometimes it makes sense to go `Value zeroIdx = ...` to make it clearer you just want the `Value`